app/lib/service/entrypoint/tools.dart

```diff
@@ -24,3 +25,25 @@ Future<void> withToolRuntime(Future<void> Function() fn) async {
     await subs.cancel();
   }
 }
+
+/// Waits for any of the listed signals and returns the first that occurs.
+Future<ProcessSignal> waitForProcessSignalTermination() {
+  final subscriptions = <StreamSubscription>[];
+  final completer = Completer<ProcessSignal>();
+
+  void process(ProcessSignal event) {
+    while (subscriptions.isNotEmpty) {
+      subscriptions.removeLast().cancel();
+    }
```

Comment:
I forget to do async `cancel`, now fixed it. I think it may be safe to call `cancel` multiple times (in case we are receiving more than one signal before exiting), but this makes sure we only cancel it once.